### PR TITLE
 Implement a single validation function around the service_uri

### PR DESCRIPTION
**Pull request by ahmedsobeh**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Do tests and lints pass with this change?
- [x] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [x] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Is there an example added to the examples folder (if applicable)?
- [ ] Was the change added to CHANGES file?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change

Created url_parser and removed duplication of parse_url. This PR resolves issue https://github.com/valkey-io/valkey-py/issues/25

I created a url_parser module and added it to the parsers folder. I then used this module in both connection modules (the normal one and the asyncio one) and updated the tests accordingly.

### CODE REVIEW

File: valkey/_parsers/url_parser.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                    raise ValueError(f"Invalid value for `{name}` in connection URL.")
+            else:
+                kwargs[name] = value
+
+    if parsed.username:
+        kwargs["username"] = unquote(parsed.username)
+    if parsed.password:
+        kwargs["password"] = unquote(parsed.password)
+
+    # We only support valkey://, valkeys:// and unix:// schemes.
```
Comment (mkmkme) [inline valkey/_parsers/url_parser.py:70]: Should we also support redis and rediss, as requested in other issue?

Comment (ahmedsobeh) [reply valkey/_parsers/url_parser.py:70]: It depends on what gets merged first 😅  If this PR gets merged first, then updating the branch of #27 with main's latest would cover it. if #27 gets merged first, I'll update my PR accordingly and it should cover it too

Comment (ahmedsobeh) [reply valkey/_parsers/url_parser.py:70]: on second thought, I'll just add it now to reduce conflicts

**Comment by aiven-sal** [comment]
LGTM, but CI is a bit cranky. Will merge later.

File: valkey/_parsers/url_parser.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        "timeout": float,
+    }
+)
+
+
+def parse_url(url: str, async_connection: bool):
+    url_prefixes = ["valkey://", "valkeys://", "unix://", "redis://", "rediss://"]
+    supported_schemes = ["valkey", "valkeys", "redis", "rediss"]
+
+    if not any(url.startswith(prefix) for prefix in url_prefixes):
```
Comment (aiven-sal) [inline valkey/_parsers/url_parser.py:42]: The exception is already raised on line 96. This `if` is redundant and it is also case-sensitive and it shouldn't be.
We can just drop the whole `if` block.

Comment (ahmedsobeh) [reply valkey/_parsers/url_parser.py]: ah that's true. removed it.

Comment (ahmedsobeh) [reply valkey/_parsers/url_parser.py]: turns out this actually covers the test **_test_invalid_scheme_raises_error_when_double_slash_missing_**. Without this, not having a double slash doesn't raise an error.

Comment (aiven-sal) [reply valkey/_parsers/url_parser.py]: Oh right. I think the best way to check for that would be something like:
```
if not parsed.netloc:
    raise something
```

Comment (ahmedsobeh) [reply valkey/_parsers/url_parser.py]: done ✅

Comment (ahmedsobeh) [reply valkey/_parsers/url_parser.py]: doesn't really work with unix, so I added a regex for it

**Review by aiven-sal — APPROVED** [review]
LGTM
I squashed and rebased some commits and will merge as soon as the CI completes

**Pull request merged**